Automattic/wp-calypso
# Checkout: Use shopping cart items to provide product variants

#### Proposed Changes

When viewing checkout, some products have billing term variants available (eg: for a Personal plan you can choose between 1 month, 1 year, or 2 years). These variants are fetched with a complicated combination of data by the `useGetProductVariants()` hook. The hook uses the product slug and quantity from the shopping cart, fetches plan and product data from two other endpoints, then combines that with hard-coded product data in calypso. The process of doing this has several downsides, notably that its prices can be wrong (https://github.com/Automattic/wp-calypso/issues/69964 and https://github.com/Automattic/wp-calypso/issues/55019) and that it requires quite a lot of frontend machinery to recreate data the backend already has.

In D93550-code, we amend the shopping-cart endpoint to include product variant data for each cart item. This PR uses that data to greatly simplify `useGetProductVariants()` so that it no longer needs the product, plan, or hard-coded information. In addition, many of the product variant guards (like not being allowed to purchase plan upgrades that are term downgrades) are no longer required as they will now be handled by the endpoint.

If it works correctly, this PR should not affect the behavior of calypso at all.

#### Screenshots

<img width="573" alt="Screenshot 2022-11-26 at 1 00 44 AM" src="https://user-images.githubusercontent.com/2036909/204074719-1024aa03-3988-4893-a005-e466abb5129f.png">

<img width="580" alt="Screenshot 2022-11-26 at 1 00 56 AM" src="https://user-images.githubusercontent.com/2036909/204074723-6ba78626-c0af-4806-882c-a3f46f9a224d.png">

<img width="595" alt="Screenshot 2022-11-26 at 1 06 58 AM" src="https://user-images.githubusercontent.com/2036909/204074872-7f57a921-125e-434d-93a2-0149309d9ef4.png">


#### Testing Instructions

To test this properly we'll need to load the shopping cart with various types of products with and without this PR to verify that the product variants are unchanged. 

Some situations to check, comparing each one against the variant picker in production:

- WordPress.com plans with and without discounts.
- Jetpack bundles with and without discounts.
- Jetpack products that have a quantity set.
